packages/go/analysis/ad/adcs.go

```diff
@@ -27,13 +27,198 @@ import (
+		if hasEku, err := certTemplateHasEkuOrAll(certTemplateOne, EkuCertRequestAgent, EkuAnyPurpose); err != nil {
+			log.Errorf("Error checking ekus for certtemplate %d: %v", certTemplateOne.ID, err)
+		} else if !hasEku {
+			continue
+		} else if domainNode, err := getDomainForCertTemplate(tx, certTemplateOne); err != nil {
+			log.Errorf("Error getting domain node for certtemplate %d: %v", certTemplateOne.ID, err)
+		} else if hasPath, err := DoesCertTemplateLinkToDomain(tx, certTemplateOne, domainNode); err != nil {
+			log.Errorf("Error fetching paths from certtemplate %d to domain: %v", certTemplateOne.ID, err)
+		} else if !hasPath {
+			continue
+		} else {
+			for _, certTemplateTwo := range versionOneCertTemplates {
+				if certTemplateTwo.ID == certTemplateOne.ID {
+					continue
+				} else if hasPath, err := DoesCertTemplateLinkToDomain(tx, certTemplateTwo, domainNode); err != nil {
+					log.Errorf("Error getting domain node for certtemplate %d: %v", certTemplateTwo.ID, err)
+				} else if !hasPath {
+					continue
+				} else {
+					outC <- analysis.CreatePostRelationshipJob{
+						FromID: 0,
+						ToID:   0,
+						Kind:   nil,
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func getDomainForCertTemplate(tx graph.Transaction, certTemplate graph.Node) (graph.Node, error) {
+	if domainSid, err := certTemplate.Properties.Get(ad.DomainSID.String()).String(); err != nil {
+		return graph.Node{}, err
+	} else if domainNode, err := analysis.FetchNodeByObjectID(tx, domainSid); err != nil {
+		return graph.Node{}, err
+	} else {
+		return *domainNode, nil
+	}
+}
+
+func enrollOnBehalfOfSelfControl(tx graph.Transaction, versionOneCertTemplates []graph.Node, outC chan<- analysis.CreatePostRelationshipJob) error {
+	for _, certTemplate := range versionOneCertTemplates {
+		if hasEku, err := certTemplateHasEkuOrAll(certTemplate, EkuAnyPurpose); err != nil {
+			log.Errorf("Error checking ekus for certtemplate %d: %v", certTemplate.ID, err)
+		} else if !hasEku {
+			continue
+		} else if subjectRequireUpn, err := certTemplate.Properties.Get(ad.SubjectAltRequireUPN.String()).Bool(); err != nil {
+			log.Errorf("Error getting subjectAltRequireUPN for certtemplate %d: %v", certTemplate.ID, err)
+		} else if !subjectRequireUpn {
+			continue
+		} else if domainNode, err := getDomainForCertTemplate(tx, certTemplate); err != nil {
+			log.Errorf("Error getting domain for certtemplate %d: %v", certTemplate.ID, err)
+		} else if doesLink, err := DoesCertTemplateLinkToDomain(tx, certTemplate, domainNode); err != nil {
+			log.Errorf("Error fetching paths from certtemplate %d to domain: %v", certTemplate.ID, err)
+		} else if !doesLink {
+			continue
+		} else {
+			outC <- analysis.CreatePostRelationshipJob{
+				FromID: certTemplate.ID,
+				ToID:   certTemplate.ID,
+				Kind:   ad.EnrollOnBehalfOf,
+			}
+		}
+	}
+
+	return nil
+}
+
+func certTemplateHasEkuOrAll(certTemplate graph.Node, targetEkus ...string) (bool, error) {
```

Comment:
Maybe these two funcs can be combined by taking an additional boolean param like `considerAll`. In the `else if` you can check `considerAll && len(ekus) == 0`.
I'll leave this up to you. It might make the logic more opaque at the cost of making it more concise.

Reply:
I went back and forth a bit on this, but I think having it separated makes it clear the intention of each one

Reply:
Sounds good 👍